File: client-libs/python/hybriddb.py
"""
HybridDB Python Client Library

Thin client - ALL database logic in C++ server
This only handles socket communication
"""

import socket
import struct
import json
from typing import List, Dict, Any, Optional

class HybridDB:
    MSG_CONNECT = 0x01
    MSG_DISCONNECT = 0x02
    MSG_QUERY = 0x03
    MSG_RESULT = 0x04
    MSG_ERROR = 0x05
    MSG_BEGIN_TXN = 0x06
    MSG_COMMIT_TXN = 0x07
    MSG_ROLLBACK_TXN = 0x08
    
    def __init__(self, host='localhost', port=5432):
        self.host = host
        self.port = port
        self.socket = None
        self.connected = False
        self.current_txn = None
        self._connect()
    
    def _connect(self):
        """Connect to C++ database server"""
        self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.socket.connect((self.host, self.port))
        self.connected = True
    
    def _send_message(self, msg_type: int, payload: bytes = b''):
        """Send message to server"""
        length = len(payload)
        message = struct.pack('!BI', msg_type, length) + payload
        self.socket.sendall(message)
    
    def _receive_message(self) -> tuple:
        """Receive message from server"""
        header = self.socket.recv(5)
        if not header:
            raise ConnectionError("Connection lost")
        
        msg_type, length = struct.unpack('!BI', header)
        payload = self.socket.recv(length)
        
        return msg_type, payload
    
    def query(self, sql: str) -> List[Dict]:
        """Execute SQL query - ALL PROCESSING ON C++ SERVER!"""
        self._send_message(self.MSG_QUERY, sql.encode('utf-8'))
        msg_type, payload = self._receive_message()
        
        if msg_type == self.MSG_ERROR:
            raise Exception(payload.decode('utf-8'))
        
        return json.loads(payload.decode('utf-8'))
    
    def begin(self) -> bool:
        """Begin transaction"""
        self._send_message(self.MSG_BEGIN_TXN)
        msg_type, payload = self._receive_message()
        
        if msg_type == self.MSG_RESULT:
            self.current_txn = json.loads(payload.decode('utf-8'))['txn_id']
            return True
        return False
    
    def commit(self) -> bool:
        """Commit transaction"""
        if not self.current_txn:
            raise Exception("No active transaction")
        
        self._send_message(self.MSG_COMMIT_TXN)
        msg_type, _ = self._receive_message()
        
        self.current_txn = None
        return msg_type == self.MSG_RESULT
    
    def rollback(self) -> bool:
        """Rollback transaction"""
        if not self.current_txn:
            raise Exception("No active transaction")
        
        self._send_message(self.MSG_ROLLBACK_TXN)
        msg_type, _ = self._receive_message()
        
        self.current_txn = None
        return msg_type == self.MSG_RESULT
    
    # Helper methods
    def create_table(self, name: str, columns: Dict[str, str]) -> Any:
        """Create table"""
        cols = ', '.join(f"{k} {v}" for k, v in columns.items())
        return self.query(f"CREATE TABLE {name} ({cols})")
    
    def insert(self, table: str, data: Dict[str, Any]) -> Any:
        """Insert row"""
        cols = ', '.join(data.keys())
        vals = ', '.join(f"'{v}'" if isinstance(v, str) else str(v) for v in data.values())
        return self.query(f"INSERT INTO {table} ({cols}) VALUES ({vals})")
    
    def select(self, table: str, where: Optional[str] = None) -> List[Dict]:
        """Select rows"""
        sql = f"SELECT * FROM {table}"
        if where:
            sql += f" WHERE {where}"
        return self.query(sql)
    
    def update(self, table: str, data: Dict[str, Any], where: str) -> Any:
        """Update rows"""
        set_clause = ', '.join(f"{k} = '{v}'" if isinstance(v, str) else f"{k} = {v}" 
                              for k, v in data.items())
        return self.query(f"UPDATE {table} SET {set_clause} WHERE {where}")
    
    def delete(self, table: str, where: str) -> Any:
        """Delete rows"""
        return self.query(f"DELETE FROM {table} WHERE {where}")
    
    def close(self):
        """Close connection"""
        if self.connected:
            self._send_message(self.MSG_DISCONNECT)
            self.socket.close()
            self.connected = False
    
    def __enter__(self):
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()
    
    def __del__(self):
        self.close()

File: client-libs/php/HybridDB.php
<?php
/**
 * HybridDB PHP Client Library
 * 
 * Thin client - ALL database logic in C++ server
 * This only handles socket communication
 */

class HybridDB {
    private $socket;
    private $host;
    private $port;
    private $connected = false;
    private $currentTxn = null;
    
    const MSG_CONNECT = 0x01;
    const MSG_DISCONNECT = 0x02;
    const MSG_QUERY = 0x03;
    const MSG_RESULT = 0x04;
    const MSG_ERROR = 0x05;
    const MSG_BEGIN_TXN = 0x06;
    const MSG_COMMIT_TXN = 0x07;
    const MSG_ROLLBACK_TXN = 0x08;
    
    public function __construct($host = 'localhost', $port = 5432) {
        $this->host = $host;
        $this->port = $port;
        $this->connect();
    }
    
    private function connect() {
        $this->socket = socket_create(AF_INET, SOCK_STREAM, SOL_TCP);
        
        if (!$this->socket) {
            throw new Exception("Failed to create socket");
        }
        
        if (!socket_connect($this->socket, $this->host, $this->port)) {
            throw new Exception("Failed to connect to {$this->host}:{$this->port}");
        }
        
        $this->connected = true;
    }
    
    private function sendMessage($type, $payload = '') {
        $length = strlen($payload);
        $message = pack('C', $type) . pack('V', $length) . $payload;
        
        socket_write($this->socket, $message, strlen($message));
    }
    
    private function receiveMessage() {
        $header = socket_read($this->socket, 5);
        if (!$header) {
            throw new Exception("Connection lost");
        }
        
        $data = unpack('Ctype/Vlength', $header);
        $payload = socket_read($this->socket, $data['length']);
        
        return [
            'type' => $data['type'],
            'payload' => $payload
        ];
    }
    
    // Execute SQL query - ALL PROCESSING ON C++ SERVER!
    public function query($sql) {
        $this->sendMessage(self::MSG_QUERY, $sql);
        $response = $this->receiveMessage();
        
        if ($response['type'] == self::MSG_ERROR) {
            throw new Exception($response['payload']);
        }
        
        return json_decode($response['payload'], true);
    }
    
    // Transaction methods
    public function begin() {
        $this->sendMessage(self::MSG_BEGIN_TXN);
        $response = $this->receiveMessage();
        
        if ($response['type'] == self::MSG_RESULT) {
            $this->currentTxn = json_decode($response['payload'], true)['txn_id'];
            return true;
        }
        return false;
    }
    
    public function commit() {
        if (!$this->currentTxn) {
            throw new Exception("No active transaction");
        }
        
        $this->sendMessage(self::MSG_COMMIT_TXN);
        $response = $this->receiveMessage();
        
        $this->currentTxn = null;
        return $response['type'] == self::MSG_RESULT;
    }
    
    public function rollback() {
        if (!$this->currentTxn) {
            throw new Exception("No active transaction");
        }
        
        $this->sendMessage(self::MSG_ROLLBACK_TXN);
        $response = $this->receiveMessage();
        
        $this->currentTxn = null;
        return $response['type'] == self::MSG_RESULT;
    }
    
    // Helper methods
    public function createTable($name, $columns) {
        $cols = implode(', ', array_map(function($col, $type) {
            return "$col $type";
        }, array_keys($columns), $columns));
        
        return $this->query("CREATE TABLE $name ($cols)");
    }
    
    public function insert($table, $data) {
        $cols = implode(', ', array_keys($data));
        $vals = implode(', ', array_map(function($v) {
            return is_string($v) ? "'$v'" : $v;
        }, $data));
        
        return $this->query("INSERT INTO $table ($cols) VALUES ($vals)");
    }
    
    public function select($table, $where = null) {
        $sql = "SELECT * FROM $table";
        if ($where) {
            $sql .= " WHERE $where";
        }
        return $this->query($sql);
    }
    
    public function update($table, $data, $where) {
        $set = implode(', ', array_map(function($k, $v) {
            return "$k = " . (is_string($v) ? "'$v'" : $v);
        }, array_keys($data), $data));
        
        return $this->query("UPDATE $table SET $set WHERE $where");
    }
    
    public function delete($table, $where) {
        return $this->query("DELETE FROM $table WHERE $where");
    }
    
    public function __destruct() {
        if ($this->connected) {
            $this->sendMessage(self::MSG_DISCONNECT);
            socket_close($this->socket);
        }
    }
}
?>